### PR TITLE
Add IndexTemplateV2 to MetaData

### PR DESCRIPTION
**Pull request by dakrone**
This adds the `IndexTemplateV2` and `IndexTemplateV2Metadata` class to be used for the new
implementation of index templates. The new metadata is stored as a `MetaData.Custom` implementation.

Relates to #53101

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateV2Metadata.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * The {@link IndexTemplateV2Metadata} class is a custom {@link MetaData.Custom} implementation that
+ * stores a map of ids to {@link IndexTemplateV2} templates.
+ */
+public class IndexTemplateV2Metadata implements MetaData.Custom {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateV2Metadata.java]: maybe also add a unit test for this class?

**Review by martijnvg — APPROVED** [review]
Left one comment. LGTM

File: server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateV2Metadata.java (hunk 1 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/metadata/IndexTemplateV2Metadata.java]: Yes good idea, I'll add that

**Comment by dakrone** [comment]
@elasticmachine test this

**Comment by dakrone** [comment]
@elasticmachine test this

**Comment by dakrone** [comment]
@elasticmachine update branch

**Pull request merged**